crates/core/tedge_write/src/bin.rs

```diff
@@ -0,0 +1,160 @@
+// TODO: force `destination_path` to be the first argument in clap
+
+use std::fs;
+use std::io;
+use std::os::unix::fs::PermissionsExt;
+
+use anyhow::bail;
+use anyhow::Context;
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use clap::Parser;
+
+/// A binary used for writing to files which `tedge` user does not have write permissions for, using
+/// sudo.
+#[derive(Debug, Clone, PartialEq, Eq, Parser)]
+#[command(about, version, long_about = None)]
+pub struct Args {
+    /// A canonical path to a file which will be written to.
+    ///
+    /// If the file does not exist, it will be created. If the file does exist and is not empty, the
+    /// file will be overridden.
+    destination_path: Utf8PathBuf,
+
+    /// Permission mode for the file, in octal form.
+    #[arg(long)]
+    mode: Option<Box<str>>,
+
+    /// User which will become the new owner of the file.
+    #[arg(long)]
+    user: Option<Box<str>>,
+
+    /// Group which will become the new owner of the file.
+    #[arg(long)]
+    group: Option<Box<str>>,
+}
+
+pub fn run(args: Args) -> anyhow::Result<()> {
+    // /etc/sudoers can contain rules where sudo permissions are given to `tedge` user depending on
+    // the files we write to, e.g:
+    //
+    // `tedge    ALL = (ALL) NOPASSWD: /usr/bin/tedge-write /etc/*`
+    //
+    // If the destination path contains `..` then we can "escape" outside the directory we're
+    // allowed to write to. For that reason, we require paths to be canonical.
+    //
+    // Ideally this would be solved by a more expressive filesystem permissions system, e.g. ACL,
+    // but unfortunately they're not standard on Linux, so we're stuck with trying to do next best
+    // thing with sudo.
+    if !args.destination_path.is_absolute() {
+        bail!("Destination path has to be absolute");
+    }
+
+    let target_filepath: Utf8PathBuf = path_clean::clean(args.destination_path.as_str()).into();
+
+    if target_filepath != *args.destination_path {
+        bail!(
+            "Destination path {} is not canonical",
+            args.destination_path
+        );
+    }
+
+    let file_existed_before_write = target_filepath.is_file();
+
+    write_stdin_to_file_atomic(&target_filepath)?;
+
+    if file_existed_before_write {
+        return Ok(());
+    }
```

Comment:
That's a good idea to avoid changing file ownership in this case.
This extends the number of cases where `tedge-write` can be used without `sudo`.